File: CN/go_back_n/sender.cpp
#include <iostream>
#include <winsock2.h>
#include <ws2tcpip.h>
#include <vector>
#include <string>
#pragma comment(lib, "ws2_32.lib")

using namespace std;

int main() {
    WSADATA wsa;
    WSAStartup(MAKEWORD(2,2), &wsa);

    SOCKET sockfd;
    char buffer[1024];
    sockaddr_in senderAddr{}, receiverAddr{};
    int addr_len = sizeof(receiverAddr);

    sockfd = socket(AF_INET, SOCK_DGRAM, 0);

    senderAddr.sin_family = AF_INET;
    senderAddr.sin_port = htons(5000);
    senderAddr.sin_addr.s_addr = inet_addr("127.0.0.1");

    bind(sockfd, (sockaddr*)&senderAddr, sizeof(senderAddr));

    receiverAddr.sin_family = AF_INET;
    receiverAddr.sin_port = htons(5001);
    receiverAddr.sin_addr.s_addr = inet_addr("127.0.0.1");

    int timeout = 2000;
    setsockopt(sockfd, SOL_SOCKET, SO_RCVTIMEO,
               (const char*)&timeout, sizeof(timeout));

    vector<string> frames = {"A","B","C","D","E","F","G","H"};

    int base = 0;
    int nextSeq = 0;
    int windowSize = 4;
    int n = frames.size();

    cout << "SENDER STARTED (GO-BACK-N)\n\n";

    while (base < n) {

        // send window
        while (nextSeq < base + windowSize && nextSeq < n) {
            string frame = "FRAME:" + to_string(nextSeq) + ":" + frames[nextSeq];

            cout << "SENDER: Sending -> " << frame << endl;

            sendto(sockfd, frame.c_str(), frame.size(), 0,
                   (sockaddr*)&receiverAddr, addr_len);

            nextSeq++;
        }

        memset(buffer, 0, sizeof(buffer));

        int bytes = recvfrom(sockfd, buffer, sizeof(buffer), 0, NULL, NULL);

        if (bytes == SOCKET_ERROR) {
            cout << "SENDER: Timeout, retransmitting window\n";

            for (int i = base; i < nextSeq; i++) {
                string frame = "FRAME:" + to_string(i) + ":" + frames[i];

                cout << "SENDER: Retransmitting -> " << frame << endl;

                sendto(sockfd, frame.c_str(), frame.size(), 0,
                       (sockaddr*)&receiverAddr, addr_len);
            }

            cout << endl;
            continue;
        }

        cout << "SENDER: Received -> " << buffer << endl;

        int ack = stoi(string(buffer).substr(4));

        if (ack >= base) {
            base = ack + 1;
            cout << "SENDER: Sliding window, new base = " << base << endl;
        }

        cout << endl;
    }

    cout << "ALL FRAMES SENT SUCCESSFULLY\n";

    closesocket(sockfd);
    WSACleanup();
    return 0;
}

File: PRAC_EXAM/CN/4_DLL/go_back_n/sender.py
import socket

HOST = "127.0.0.1"
PORT = 5100

sock = socket.socket()
sock.connect((HOST, PORT))
sock.settimeout(2)

frames = ["A", "B", "C", "D", "E", "F", "G", "H"]

base = 0
next_seq = 0
window_size = 4

print("SENDER STARTED\n")

while base < len(frames):

    # Send frames in the current window
    while next_seq < base + window_size and next_seq < len(frames):
        frame = f"FRAME:{next_seq}:{frames[next_seq]}"
        print("Sending:", frame)
        sock.send(frame.encode())
        next_seq += 1

    try:
        msg = sock.recv(1024).decode()
        ack = int(msg.split(":")[1])

        print("Received:", msg)

        if ack >= base:
            base = ack + 1

    except socket.timeout:
        print("Timeout! Retransmitting window...")

        for i in range(base, next_seq):
            frame = f"FRAME:{i}:{frames[i]}"
            print("Retransmitting:", frame)
            sock.send(frame.encode())

print("\nAll frames sent successfully.")
sock.close()